lib/CIndexStoreDB/CIndexStoreDB.cpp

```diff
@@ -144,6 +144,37 @@ indexstoredb_symbol_name(indexstoredb_symbol_t symbol) {
   return obj->value->getName().c_str();
 }
 
+/// loops through each symbol in the index and calls the receiver function with each symbol
+/// @param index an IndexStoreDB object which contains the symbols
+/// @param receiver a function to be called for each symbol, the CString of the symbol will be passed in to this function.
+/// The function should return a boolean indicating whether the looping should continue.
+bool
+indexstoredb_for_each_symbol_name(indexstoredb_index_t index, indexstoredb_for_each_symbol_receiver receiver) {
+  // indexSystem has foreachsymbolName.
+  auto obj = (IndexStoreDBObject<std::shared_ptr<IndexSystem>> *)index;
+  return obj->value->foreachSymbolName([&](StringRef ref) -> bool {
+    return receiver(ref.str().c_str());
```

Comment:
The number of symbol names is going to be huge in large projects, and this is copying every name string to a `std::string`. I recommend to eliminate the string copy by:
* Have the name receiver receive the StringRef pointer and size
* On the Swift side, construct a String by forming a `UnsafeBufferPointer<UInt8>` from the pointer+size pair and pass it to `String(decoding: textBuffer, as: UTF8.self)`

Reply:
oh, thank you! I was hoping there was a way to make this better. Your other comments are also helpful. I'll work on this as soon as I get a chance.

Reply:
Actually, I don't think I'm going to need to worry about the copying problem anymore if I use the pattern matching method. Is that correct?

Reply:
It will be general goodness to expose the other functions as well, since we may need them later on in the future, but it is not a requirement, feel free to just expose the pattern function if you'd like to keep it simple.

Reply:
Let's not expose this function if we're not going to fix the string performance issue.  We can always come back to it later.